Implementations of platform agnostic portions of FML.

chinmaygarde opened the pull request:
This patch does not include updates to the BUILD.gn files or the platform specific backends. This allows us to land these patches without breaking existing builds. Meanwhile, the platform specific backends will be sent in different PRs for easier review.

This patch builds upon the suggestions made on https://github.com/flutter/engine/pull/3459 (which was abandoned). All comments on that PR have been addressed here.

abarth commented on fml/icu_util.cc:
```diff
@@ -0,0 +1,91 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/icu_util.h"
```
blank line after primary header

chinmaygarde replied:
Will fix all files.

abarth commented on fml/icu_util.cc:
```diff
@@ -0,0 +1,91 @@
+
+#include "flutter/fml/icu_util.h"
+#include <memory>
+#include <mutex>
```
blank line after system headers

abarth commented on fml/icu_util.cc:
```diff
@@ -0,0 +1,91 @@
+namespace fml {
+namespace icu {
+
+static const char kIcuDataFileName[] = "icudtl.dat";
```
constexpr

chinmaygarde replied:
Done.

abarth commented on fml/icu_util.cc:
```diff
@@ -0,0 +1,91 @@
+  FTL_CHECK(context->IsValid()) << "Must be able to initialize the ICU context";
+}
+
+std::once_flag gICUInitFlag;
```
g_icu_init_flag

chinmaygarde replied:
Done.

abarth commented on fml/icu_util.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef FLUTTER_FML_ICU_UTIL_H_
+#define FLUTTER_FML_ICU_UTIL_H_
+
+#include <string>
```
blank line after system headers

abarth commented on fml/mapping.cc:
```diff
@@ -0,0 +1,84 @@
+#include <unistd.h>
+#include <type_traits>
+#include "lib/ftl/build_config.h"
+#include "lib/ftl/files/eintr_wrapper.h"
```
header madness

chinmaygarde replied:
I don't understand what can be done.

abarth replied:
I just meant to add blank lines.  :)

abarth commented on fml/message_loop.cc:
```diff
@@ -0,0 +1,68 @@
+
+namespace fml {
+
+FML_THREAD_LOCAL ThreadLocal TLSMessageLoop([](intptr_t value) {
```
unix_hacker_names_please

chinmaygarde replied:
Done.

abarth commented on fml/message_loop.cc:
```diff
@@ -0,0 +1,68 @@
+
+MessageLoop& MessageLoop::GetCurrent() {
+  auto loop = reinterpret_cast<MessageLoop*>(TLSMessageLoop.Get());
+  FTL_DCHECK(loop != nullptr)
```
If you're going to return a reference, then this should be an FTL_CHECK otherwise you'll get undefined behavior.

chinmaygarde replied:
Done. Converted to FTL_CHECK.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

abarth commented:
Looks great.

Hixie commented:
Side note: we may wish to rename "FML" to something with an acronym that doesn't coincide with a popular informal acronym. We may eventually regret the name when doing public talks about it.

chinmaygarde commented:
Addressed all comments and updated licenses.

> we may wish to rename "FML"

None of the interfaces are public and we own the library. So we can safely change this in the future if needed.
